Update GPU technote for 1.31; doc atomic functions in GPU module (#22556)
chapel-lang/chapel: doc/rst/technotes/gpu.rst

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

To enable GPU support set the environment variable ``CHPL_LOCALE_MODEL=gpu``
before building Chapel. Several other variables affect how Chapel generates
code for and interacts with GPUs. These variables include:

> * ``CHPL_GPU`` --- may be set to ``nvidia``, ``amd', or ``cpu``. If unset, as

Reviewer: ```
``amd``
```
Perhaps?

  part of its build process, Chapel will attempt to automatically determine
  what type of GPU you're trying to target. Changing this variable requires

>   rebuilding Chapel. For more information, see the `Vendor Portability`_

Reviewer: All of Chapel? I thought it was just the runtime, and it may not even require a rebuild. I mean, I thought Engin closed that issue I opened...

https://github.com/chapel-lang/chapel/issues/21978

Author: Great point. Reworded to "may require rebuilding the Chapel runtime" ("may" since if you switch back-and-forth I don't think it would require rebuilding again).

  section.
* ``CHPL_GPU_ARCH`` --- specifies GPU architecture to generate kernel code for.

>   If unset and targeting NVIDIA GPUs, will default to 'sm_60'. This may also

Reviewer: ```
``sm_60``
```
?

  be set by passing the ``chpl`` compiler ``--gpu-arch=<architecture>``. For
  more information, see the `Vendor Portability`_ section.

> * ``CHPL_CUDA_PATH`` --- specifies path to CUDA toolkit.  If unset, Chapel tries

Reviewer: Nitpick, but "path _to_"?

  to automatically determine this path based on the location of ``nvcc``. This
  variable is unused if not targeting NVIDIA GPUs. For more information, see
  the `Vendor Portability`_ section.
* ``CHPL_ROCM_PATH`` --- specifies path to ROCm library. If unset, Chapel tries
  to automatically determine this path based on the location of ``hipcc``. This
  variable is unused if not targeting AMD GPUs. For more information, see the
  `Vendor Portability`_ section.

> * ``CHPL_RT_NUM_GPU_PER_LOCALE`` --- if using ``CHPL_GPU=cpu``, sets how many

Reviewer: It was my understanding that if not using `CHPL_GPU=cpu`, if a locale had N GPUs, and you set this value to K < N, only the first K GPUs would be used.

Author: > It was my understanding that if not using CHPL_GPU=cpu, if a locale had N GPUs, and you set this value to K < N, only the first K GPUs would be used.

You're right, How's this:

```
* ``CHPL_RT_NUM_GPUS_PER_LOCALE`` --- sets how many GPU sublocales to have per
  locale. If using ``CHPL_GPU=cpu``, may be set to any non negative value,
  otherwise it may be set to any value equal-to or lower than the number of
  GPUs available on each node.  If unset, defaults to 1 if ``CHPL_GPU=cpu``,
  otherwise defaults to the number of GPUs available on each node.
  For more information, see the `CPU as Device mode`_ section.
```

I suppose it's really "f using ``CHPL_GPU=cpu``, may be set to any non negative value that fits in a signed 32-bit integer" but I don't think that's worth spelling out.

Reviewer: How about:
> If unset, defaults to the number of GPUs available on each node, except for `CHPL_GPU=cpu`, in which case it defaults to 1.

  Device mode`_ section.
* ``CHPL_GPU_MEM_STRATEGY`` --- dictates how to allocate data when on a GPU
  locale.  May be set to ``unified_memory`` or ``array_on_device``. If unset,
  defaults to ``unified_memory``. Changing this variable requires rebuilding
  Chapel. For more information, see the `Memory Strategies`_ section.
* ``CHPL_GPU_BLOCK_SIZE`` --- specifies default block size when launching
  kernels. If unset, defaults to 512. This variable may also be set by passing
  the ``chpl`` compiler ``--gpu-block-size=<block_size>``. It can also be
  overwritten on a per-kernel basis by using the :proc:`~GPU.setBlockSize`
  function.
* ``CHPL_GPU_SPECIALIZATION`` --- if set, outlines bodies of 'on' statements
  and clones all functions reachable from that block. The 'on' statement is
  rewritten to call the cloned version of the outlined function when on a  GPU
  locale. This mode increases overall code size but allows the compiler to
  assume that a given function will execute on the GPU locale and optimize
  accordingly. This may also be set by passing the ``chpl`` compiler the
  ``--gpu-specialization`` flag. This is an experimental mode subject to change
  in the future.

> * ``CHPL_GPU_NO_CPU_MODE_WARNING`` - this variable is relevant when using the
>   `CPU as Device mode`_ and if set causes it so that uses of
>   :proc:`~GPU.assertOnGpu` while in this mode will not generate a warning at
>   execution time. Alternatively, this can be done by passing
>   ``--gpuNoCpuModeWarning`` to your application. For more information, see the
>   `CPU as Device mode`_ section.

Reviewer: ```suggestion
* ``CHPL_GPU_NO_CPU_MODE_WARNING`` - this variable is relevant when using the
  `CPU as Device mode`_ and if set causes uses of
  :proc:`~GPU.assertOnGpu` to not generate a warning at
  execution time. Alternatively, this behavior can be enabled by passing
  ``--gpuNoCpuModeWarning`` to your application. For more information, see the
  `CPU as Device mode`_ section.
```
Forgive my line breaking.

Author: Huh, I wasn't aware GH would let you post suggestions like that with a "Commit suggestion" button to just apply it.

I'm not going to hit it because I'm updating things in my editor (and I'm not sure if the DCO line for my commit message would get through), but that's pretty cool.

Features
--------------------

In the following subsections we discuss various features of GPU supports.

Vendor Portability
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Chapel is able to generate code that will execute on either NVIDIA or AMD GPUs.
Chapel’s build system will automatically try and deduce what type of GPU you
have and where your installation of relevant runtime (e.g. CUDA or ROCm) are.
If the type of GPU is not detected you may set the ``CHPL_GPU`` environment
variable manually to either ``nvidia`` or ``amd``.  ``CHPL_GPU`` may also
manually be set to ``cpu`` to use `CPU as Device mode`_.

Based on the value of ``CHPL_GPU``, Chapel's build system will also attempt to
automatically detect the path to the relevant runtime. If it is not
automatically detected (or you would like to use a different installation) you
may set ``CHPL_CUDA_PATH`` and/or ``CHPL_ROCM_PATH`` explicitly.

> The CHPL_GPU_ARCH environment variable can be set to control the desired GPU

Reviewer: ```
``CHPL_GPU_ARCH``
```

architecture to compile for. The default value is ``sm_60`` for
``CHPL_GPU_CODEGEN=cuda``. You may also use the ``--gpu-arch`` compiler flag to

> set GPU architecture. For a list of possible values please refer to "processor"

Reviewer: It would be nice if you could formulate it such that "for AMD" came before the link; a reader like me might click the link before realizing the table isn't for CUDA.

Author: Modified to:

```
The ``CHPL_GPU_ARCH`` environment variable can be set to control the desired
GPU architecture to compile for. The default value is ``sm_60`` for
``CHPL_GPU_CODEGEN=cuda``. You may also use the ``--gpu-arch`` compiler flag to
set GPU architecture.  If using AMD, `this table in the LLVM documentation
<https://llvm.org/docs/AMDGPUUsage.html#processors>`_ has possible architecture
values (see the "processor" column). For NVIDIA, see the `CUDA Programming
Guide
<https://docs.nvidia.com/cuda/cuda-c-programming-guide/#features-and-technical-specifications>`_.
```

values in `this table in the LLVM documentation
<https://llvm.org/docs/AMDGPUUsage.html#processors>`_ for AMD or the `CUDA
Programming Guide
<https://docs.nvidia.com/cuda/cuda-c-programming-guide/#features-and-technical-specifications>`_
for NVIDIA.

CPU as Device Mode
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
The ``CHPL_GPU`` environment variable can be set to ``cpu`` to enable many GPU
features to be used without requiring any GPUs and/or vendor SDKs to be
installed. This mode is mainly for initial development steps or quick feature
tests where access to GPUs may be limited. In this mode:

> * the compiler will generate GPU kernels from order-independent loops normally,

Reviewer: No other subsequent bullets have commas. Generally (nitpick), all the bullets here have inconsistent punctuation.

Author: This is a section Engin wrote that I moved, but I agree we should try for a consistent style within the list and more broadly within the document.

Turning a bulleted list into a kind of long sentence with things comma'd together is something I've done/seen in the past on academic papers and the like.  So I'm not really sure what's considered the best practice stylistically.

But anyway, for now I'm going to turn each bullet into a separate sentence given that's what we do elsewhere.

* it will call the internal runtime API for GPU operations, so that features
  outlined under `Diagnostics and Utilities`_ will work as expected

  * e.g, :proc:`~GPU.assertOnGpu` will fail at compile time normally. This can
    allow testing if a loop is GPU-eligible.

  * but it will generate a warning per-iteration at execution time.

  * ``CHPL_GPU_NO_CPU_MODE_WARNING`` environment can be set to suppress these
    warnings. Alternatively, you can pass ``--gpuNoCpuModeWarning`` to your
    application to the same effect.

* even though the GPU diagnostics are collected, the loop will be executed for
  correctness testing and there will not be any kernel launch

* advanced features like ``syncThreads`` and ``createSharedArray`` will compile
  and run, but in all likelihood code that uses those features will not
  generate correct results

> * ``asyncGpuComm`` will do a blocking memcpy and ``gpuCommWait`` will return

Reviewer: ```
``memcpy``
```
?
